### PR TITLE
client/tso: fix the bug that TSO may hang when switching mode

### PR DESCRIPTION
**Pull request by JmPotato**
### What problem does this PR solve?

Issue Number: ref #7849.

### What is changed and how does it work?

```commit-message
Because the old `(*tsoClient).dispatchRequest` only checked the request's context,
this could potentially cause a `tsoRequest` to be sent to a batch channel that has already
finished `revokePendingTokenRequest`, thus being left there forever without returning.
```

### Check List

Tests

- Unit test
- Integration test

### Release note

```release-note
None.
```

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -782,23 +782,39 @@ func (c *client) GetLocalTSAsync(ctx context.Context, dcLocation string) TSFutur
...
+		tsoClient := c.getTSOClient()
+		if tsoClient == nil {
+			err = errs.ErrClientGetTSO.FastGenByArgs("tso client is nil")
+			continue
 		}
+		retryable, err = tsoClient.dispatchRequest(req)
+		if !retryable {
+			break
+		}
+		time.Sleep(dispatchRetryDelay)
```
Comment (CabinfeverB) [inline client/client.go:815]: No need wait for the last retry

**Review by CabinfeverB — COMMENTED** [review]
rest LGTM

File: client/client.go [hunk 2]
```diff
@@ -782,23 +782,39 @@ func (c *client) GetLocalTSAsync(ctx context.Context, dcLocation string) TSFutur
...
+	var (
+		retryable bool
+		err       error
+	)
+	for i := 0; i < dispatchRetryCount; i++ {
+		// Get the tsoClient for each retry, as it may be initialized or switched during the process.
+		tsoClient := c.getTSOClient()
+		if tsoClient == nil {
+			err = errs.ErrClientGetTSO.FastGenByArgs("tso client is nil")
+			continue
```
Comment (CabinfeverB) [inline client/client.go:809]: And do we need to wait some time for this case？
Maybe we can wait in L805 except first time

**Review by CabinfeverB — APPROVED** [review]

**Review by HuSharp — APPROVED** [review]

**Pull request merged**